File: day3/fibonacci.py
def fibonacciIterative(start, end):
    fib = []

    for i in range(end + 1):
        if i == 0:
            fib.append(0)
        elif i == 1:
            fib.append(1)
        else:
            fib.append(fib[i - 1] + fib[i - 2])

    return fib[start:end + 1]

def fibonacci(n):
    if n == 0:
        return 0
    elif n == 1:
        return 1
    return fibonacci(n - 1) + fibonacci(n - 2)

def fibonacciRecursive(start, end):
    return [fibonacci(i) for i in range(start, end + 1)]

start = int(input("Enter start pos: "))
end = int(input("Enter end pos: "))

if(start == 0 or end == 0 or start > end):
    print("Invalid Position")
else:
    res1 = fibonacciIterative(start - 1, end - 1)
    res2 = fibonacciRecursive(start - 1, end - 1)

    print(f"By Iterative method: {res1}")
    print(f"By Recursive method: {res2}")

File: day5/fibonacci.r
fibonacci_iterative_range <- function(start, end) {
  if (start <= 0 || end <= 0 || start > end) {
    stop("Invalid range")
  }

  if (end == 1) {
    return(1)
  }

  fib <- numeric(end)
  fib[1] <- 1
  fib[2] <- 1

  if (end >= 3) {
    for (i in 3:end) {
      fib[i] <- fib[i - 1] + fib[i - 2]
    }
  }

  fib[start:end]
}

fib_recursive <- function(n) {
  if (n == 1 || n == 2) return(1)
  fib_recursive(n - 1) + fib_recursive(n - 2)
}

fibonacci_recursive_range <- function(start, end) {
  if (start <= 0 || end <= 0 || start > end) {
    stop("Invalid range")
  }

  result <- numeric(end - start + 1)
  idx <- 1
  for (i in start:end) {
    result[idx] <- fib_recursive(i)
    idx <- idx + 1
  }

  result
}

start <- as.integer(readline("Enter starting pos.: "))
end <- as.integer(readline("Enter ending pos.: "))
cat("By Iterative Method =", fibonacci_iterative_range(start, end), "\n")
cat("By Recursive Method =", fibonacci_recursive_range(start, end), "\n")